### ISSUE TITLE
Auto-generate Deadline (Order for Filing Fee)

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8278]
As a clerk, so that I don't have to manually create a deadline when an Order for Filing Fee is docketed, I need DAWSON to create a deadline on the case when an Order for Filing Fee (OF) is served.

When the filing fee has not been paid by a petitioner, the petitions or docket clerk manually adds a deadline to the case indicating the due date of the filing fee. Currently this must be done after docketing and serving the OF, requiring the clerk to remember to go back and create the deadline. To streamline this process, we would like DAWSON to automatically create the deadline when serving this Order.

## Acceptance Criteria
* When an OF is served, automatically create a Deadline in DAWSON attached to the case for which the OF is served.
* The Deadline:
  - Has a Due Date of the same date entered in the Date field on the Add Docket Entry screen
  - Has a Description of "Filing Fee Due"
  - Behaves in all other respects as if manually created by a user and can be interacted with in the usual manner (edit/delete functions for authorized users)

## Tasks
- [x] Add case deadline on Add Docket Entry of OF - mob https://github.com/ustaxcourt/ef-cms/pull/2785
  - [x] case deadline due date populated by form.date (ask holly why not populate the deadline based on date in order?) https://github.com/ustaxcourt/ef-cms/pull/2785
  - [x] case deadline description populated with "Filing Fee Due" https://github.com/ustaxcourt/ef-cms/pull/2785
  - [x] verify edit/delete on deadline behaves as expected https://github.com/ustaxcourt/ef-cms/pull/2785
- [x] Add case deadline on Serve (from do viewer) of OF https://github.com/ustaxcourt/ef-cms/pull/2785 
- [x] DOD 1 (https://github.com/ustaxcourt/ef-cms/pull/2791, https://github.com/flexion/ef-cms/pull/9711, https://github.com/flexion/ef-cms/pull/9712)

## Story Definition of Ready (Created on 9/26/21)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ]  Be immediately actionable for the development team
- [ ]  Be focused on users
- [ ]  Include a narrative. “As a ____, so that I can ______, I need _____.”
- [ ]  Include user-centric acceptance criteria that expresses the users' needs
- [ ]  Include user-centric testing scenarios
- [ ]  Include preconditions (story dependencies)
- [ ]  Not include design or technical solutions
- [ ]  Include non-standard testing, deployment, documentation, browser, mobile or security requirements, as well as any requirements for 3rd-party integrations

Process: Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 5-19-22)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [x] Automated test scripts have been written, including visual tests for newly added PDFs
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Acceptance criteria for the story has been met
 - [x] If there are special instructions in order to deploy into the next environment, add them as a comment in the story
 - [x] Reviewed by UX on a deployed environment.
 - [x] Reviewed by PO on a deployed environment. Can be deployed to the Court's `test` environment if prod-like data is required. Otherwise deployed to any `experimental` environment.
 - [ ] Deployed to the Court's `staging` environment.

**Comment by ttlenard** [comment]
## Test Cases

#### 1) Petitions Clerk Serves a petition to the IRS; Order for Filing Fee automatically generated in Drafts Folder; Adds OF to Docket Record; Deadline automatically created on the case.

- Log in as a Petitions Clerk
- Click on Document QC
- Click Start a Petition
- Fill out the required information and upload a document
- Make sure that the Not Paid radio button is selected, and the Order for Filing fee checkbox is checked.
- Click Create Case, then click Serve to IRS, an confirm by clicking Yes, Serve.
- Click done on the paper service page.
- Click on the Served tab from the My Document QC tab.
- Click on the docket number of the Petition you just served
- Click on the Drafts tab
- Click on the Draft Order for Filing fee
- Click on the Apply Signature link, apply the signature and click Save Signature
- Click Add Docket Entry
- Input the Date that filing fee is due (this is also indicated in the Auto-Generated Order)
- Click Save and Serve, and then Yes, Serve.
- Click Done on the Print for Paper Service screen.

Expected Results:
- [x] Tracked items tab has an exclamation point icon indicating that there is a tracked item
- [x] Blocked label appears on the Case header
- [x] Upon clicking on the tracked items tab, there is a new Deadline that was automatically added
- [x] Due date of deadline matches the date that you entered when adding the docket entry
- [x] Description states "Filing Fee Due"
- [x] Edit/Delete links are available to the Clerk




#### 2) Petitions Clerk creates an Order for Filing Fee; Adds OF to Docket Record; Deadline automatically created on the case.

- Log in as a Petitions Clerk
- Navigate to a case
- Click on the Create Button
- Select Order or Notice
- Select Order, and then for the Order Title, type "Order for Filing Fee"
- Click Continue
- Add in some Order Text, and click Save
- Apply signature, and then click Save Signature
- Click Add Docket Entry
- For Document Type, select Order for Filing Fee
- Input a date
- Click Save and Serve, and then Yes, Serve.
- Click Done on the Print for Paper Service screen (if Petitioner has paper service).

Expected Results:
- [x] Tracked items tab has an exclamation point icon indicating that there is a tracked item
- [x] Blocked label appears on the Case header
- [x] Upon clicking on the tracked items tab, there is a new Deadline that was automatically added
- [x] Due date of deadline matches the date that you entered when adding the docket entry
- [x] Description states "Filing Fee Due"
- [x] Edit/Delete links are available to the Clerk



#### 3) Docket Clerk creates an Order for Filing Fee; Adds OF to Docket Record; Deadline automatically created on the case.

- Log in as a Docket Clerk
- Navigate to a case
- Click on the Create Button
- Select Order or Notice
- Select Order, and then for the Order Title, type "Order for Filing Fee"
- Click Continue
- Add in some Order Text, and click Save
- Apply signature, and then click Save Signature
- Click Add Docket Entry
- For Document Type, select Order for Filing Fee
- Input a date
- Click Save and Serve, and then Yes, Serve.
- Click Done on the Print for Paper Service screen (if Petitioner has paper service).

Expected Results:
- [x] Tracked items tab has an exclamation point icon indicating that there is a tracked item
- [x] Upon clicking on the tracked items tab, there is a new Deadline that was automatically added
- [x] Due date of deadline matches the date that you entered when adding the docket entry
- [x] Description states "Filing Fee Due"
- [x] Edit/Delete links are available to the Clerk



#### 4) Clerk can Edit the auto-generated deadline, just as if it was a manually created deadline.

- Continuing from the previous tests
- Navigate to one of the auto-generated deadlines on a case, within the tracked items tab
- Click on the Edit link
- Edit the Due Date/description
- Click Save

Expected Results:
- [x] Edits to the Deadline are saved and appear as they normally would



#### 5) Clerk can Delete the auto-generated deadline, just as if it was a manually created deadline.

- Continuing from the previous tests
- Navigate to one of the auto-generated deadlines on a case, within the tracked items tab
- Click on the Delete link
- Confirm deletion

Expected Results:
- [x] The Deadline is deleted.
- [x] Blocked label is no longer on the case header (if this was the only item blocking the case)



#### 6) Court User runs a Deadlines report; Automatically generated deadlines for the Order for Filing Fee are listed in the report

- Log in as a court user
- Click on Reports
- Select Deadlines Report
- Input the date that the deadlines that were created in tests 1-3 are due
- Click Show Deadlines

Expected Results:
- [x] The auto-generated deadlines appear on the report
- [x] The description indicates "Filing Fee Due"

**Comment by swongCO** [comment]
@ttlenard In test case 1, you didn't specify which radio btn to select under Fee Paid. When we selected Paid, the checkbox for Order for Filing Fee didn't display and no automated deadline was generated. When we selected Not paid, then the checkbox displayed and the automated deadline was generated. Assuming the latter is the one you meant, it passes this test case. See screenshots:
![Screen Shot 2022-11-11 at 11.54.46 AM.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/679f1139-6c87-4ae6-a6ea-2f2d6bd81f68)
![Screen Shot 2022-11-11 at 11.54.54 AM.png](https://images.zenhubusercontent.com/61a7bbb7fda36d7b28c1ab23/40e0b577-cef3-4b7f-a47a-3ae13951e9ef)
